gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageBidiReadChannel.java

```diff
@@ -1,106 +1,178 @@
+  private final StorageResourceId resourceId;
+
+  private final Duration readTimeout;
+  private long position = 0;
 
   public GoogleCloudStorageBidiReadChannel(
       Storage storage,
       GoogleCloudStorageItemInfo itemInfo,
       GoogleCloudStorageReadOptions readOptions,
       ExecutorService boundedThreadPool)
       throws IOException {
-    this.storage = storage;
+    this.readTimeout = readOptions.getGrpcReadTimeout();
     this.resourceId =
         new StorageResourceId(
             itemInfo.getBucketName(), itemInfo.getObjectName(), itemInfo.getContentGeneration());
-    this.blobId =
+    BlobId blobId =
         BlobId.of(
             resourceId.getBucketName(), resourceId.getObjectName(), resourceId.getGenerationId());
-    this.readOptions = readOptions;
     this.blobReadSession =
         initializeBlobReadSession(storage, blobId, readOptions.getBidiClientTimeout());
     this.boundedThreadPool = boundedThreadPool;
+    initMetadata(itemInfo.getContentEncoding(), itemInfo.getSize());
   }
 
   private static BlobReadSession initializeBlobReadSession(
       Storage storage, BlobId blobId, int clientTimeout) throws IOException {
     try {
       return storage.blobReadSession(blobId).get(clientTimeout, TimeUnit.SECONDS);
     } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      GoogleCloudStorageEventBus.postOnException();
       throw new IOException(e);
     }
   }
 
   @Override
   public int read(ByteBuffer dst) throws IOException {
-    // TODO(dhritichopra) Add read flow
-    return 0;
-  }
+    throwIfNotOpen();
+    if (!dst.hasRemaining()) {
+      return 0;
+    }
 
-  @Override
-  public int write(ByteBuffer src) throws IOException {
-    throw new UnsupportedOperationException("Cannot mutate read-only channel");
+    if (position >= objectSize) {
+      return -1;
+    }
+
+    logger.atFinest().log(
+        "Reading up to %d bytes at position %d from '%s'", dst.remaining(), position, resourceId);
+
+    long bytesToRequest = Math.min(dst.remaining(), objectSize - position);
+
+    try {
+      ApiFuture<DisposableByteString> futureBytes =
+          blobReadSession.readAs(
+              ReadProjectionConfigs.asFutureByteString()
+                  .withRangeSpec(RangeSpec.of(position, bytesToRequest)));
+
+      DisposableByteString disposableByteString =
+          futureBytes.get(readTimeout.toNanos(), TimeUnit.NANOSECONDS);
+
+      int bytesRead;
+      try (DisposableByteString dbs = disposableByteString) {
+        ByteString byteString = dbs.byteString();
+        bytesRead = byteString.size();
+
+        if (bytesRead == 0 && position < objectSize) {
+          throw new IOException(
+              String.format("Read 0 bytes without blocking from object: '%s'", resourceId));
+        }
+
+        if (bytesRead > 0) {
+          byteString.copyTo(dst);
+          position += bytesRead;
+        }
+      }
+
+      return bytesRead > 0 ? bytesRead : -1;
```

Comment:
Can you replace -1 with a Integer variable with an appropriate name like "Unread Bytes" so that its more readable. Replace -1 everywhere in the file and test so that its clear you are evaluating this condition.